### PR TITLE
Add ability to send notifications over legacy protocol

### PR DESCRIPTION
**Pull request by altonen**
If legacy protocol was negotiated during handshake, save a sink for that protocol as well and provide the possibility of sending notifications using the legacy protocol name. The protocol is expected to keep track of which peers are using the legacy protocol and which are not and handle the incoming message correctly based on the peer record which states the used protocol version.

Needed for Polkadot's new validation protocol.

### CODE REVIEW

File: client/network/test/src/service.rs [hunk 1]
```diff
@@ -626,6 +626,111 @@ async fn fallback_name_working() {
...
+				);
+				node1.write_notification(
+					remote,
+					negotiated_fallback.unwrap(),
+					"hello over old name".as_bytes().to_vec(),
+				);
+			},
+			Event::NotificationsReceived { messages, .. } => {
+				assert_eq!(messages.len(), 1);
+				assert_eq!(messages[0].0, NEW_PROTOCOL_NAME.into());
```
Comment (altonen) [inline client/network/test/src/service.rs:723]: @sandreim 

I checked the Polkadot code and it looked like it doesn't check the protocol version of `NotificationsReceived` so notice that it uses the main protocol name for these. Changing this to use legacy protocol would require checking it for every protocol in Substrate and Polkadot which I would rather not do unless it's absolutely necessary as this is one of them temporary hacks again.

Comment (sandreim) [reply client/network/test/src/service.rs:723]: I think we do the check, at least in my PR https://github.com/paritytech/polkadot/pull/6782/files#diff-64aa50a62fe0bddebd18c842650162a2b1b86b0abda390cc325658bb07d5e222R448 the notifications are handled differently based on the protocol version.

Comment (altonen) [reply client/network/test/src/service.rs:723]: But [here](https://github.com/paritytech/polkadot/pull/6782/files#diff-64aa50a62fe0bddebd18c842650162a2b1b86b0abda390cc325658bb07d5e222R351-R353) you're populating `expected_versions` using `PeerId` and it looks like `protocol` (from which `version` is extracted) of `NotificationsReceived` is only used for printing. So if I understood the code correctly and you keep it as is (the protocol version is assigned for `NotificationSteamOpened` and stored for each peer) and you don't use the version of the `NoticationsReceived`, we don't have to change the `NotificationsReceived` handler of every protocol in Substrate/Polkadot.

Comment (sandreim) [reply client/network/test/src/service.rs:723]: yes, you are right, but the version is stored for each peer on notification stream opened (negotiated version) and is kept by each subsystem until peer disconnects.

Comment (rphmeier) [reply client/network/test/src/service.rs:723]: I remember when initially writing the upgrade code I asked @tomaka for comment and he said the intention was for notifications to be tagged with the "main" protocol ID even if there was a negotiated fallback. So there should be no need to check for legacy protocol names on individual messages, only when establishing the connection.

Comment (tomaka) [reply client/network/test/src/service.rs:723]: Yes, this fallback system is only relevant for the communication on the wire. When it comes to identifying the protocol through the API, only the main protocol is ever used.

In other words, the main protocol is used both on the wire and as an identifier for that protocol, while the fallbacks are only ever sent on the wire.

**Comment by sandreim** [comment]
FWIW, this fixes the v1/v2 compatiblity issues in https://github.com/paritytech/polkadot/pull/6782

File: client/network/test/src/service.rs (hunk 1 above)
Comment (dmitry-markin) [reply client/network/test/src/service.rs:723]: May be I don't understand something, but if the protocol keeps track of what peers are on what protocol version (after initial connection was established), why can't it use the main protocol name for sending notifications?

**Review by dmitry-markin — APPROVED** [review]
Looks good with the exception that I'm unsure if we should merge this. As stated in https://github.com/paritytech/substrate/pull/14324/files#r1223124535, the protocols in Substrate are identified by their main name, even if the fallback name was negotiated on the wire. So, may be, we should keep the version tracking inside the protocol implementation, and always send the notifications using the main protocol name.

File: client/network/test/src/service.rs (hunk 1 above)
Comment (sandreim) [reply client/network/test/src/service.rs:723]: I haven't tried this yet, I don't think there should be a problem after all ...

Comment (altonen) [reply client/network/test/src/service.rs:723]: > May be I don't understand something, but if the protocol keeps track of what peers are on what protocol version (after initial connection was established), why can't it use the main protocol name for sending notifications?

We discussed with @sandreim in DMs and IIRC there was some technical limitation or something that made this less than ideal. But if we can use the main protocol, that'd be better and I can close this PR

Comment (sandreim) [reply client/network/test/src/service.rs:723]: I think I understood your proposal wrong @altonen, but now what you meant makes sense.

**Comment by sandreim** [comment]
Using main protocol works as expected, thanks for looking into this!

**Pull request closed without merging**